### PR TITLE
fetch fee history

### LINKED ISSUE #522: 24fee calculation based on existing data
allow frontend. to. keep a list of when we change fee
use the correct fee percentage based on the event timestamp.

calculate the fee from every conversion event and sum the fee together to show 24h fees.

### PR DESCRIPTION
**Pull request by H34D**
fixes #522

### CODE REVIEW

**Comment by H34D** [comment]
@Velua another half done task

i have the following issues:

- limits on alchemy on localhost in general -> cannot get the data i need right now so falling back to the metamask provider which will work for now
- 2k block limit when fetching events over alchemy (see #523), we would need to send requests in 2k block batches this is still pending to implement
- i get loads of error messages from other parts of the application because of my changes (vue.runtime.esm.js?2b0e:619 [Vue warn]: Error in render: "TypeError: Cannot read property 'id' of undefined" ethBancor.ts?5460:5714)) it is on this line: `const viewToken = tokens.find(x => compareString(x.id, fee.id))!;` I think it has something to do with tokens not being fully loaded or any other race condition

**Comment by H34D** [comment]
results of today:

- i still get limit errors on localhost with alchemy errors.js?490f:28 Uncaught (in promise) Error: Returned error: Your app has exceeded its concurrent requests capacity.  If you have retries enabled, you can safely ignore this message. If not, check out https://docs.alchemyapi.io/documentation/rate-limits. Feel free to ask further questions in the community discord channel: https://discord.com/invite/A39JVCM. Have a great day!
- i refactored all the code to make the fee history part of the store itself
- i still get random errors of `TypeError: Cannot read property 'id' of undefined" ethBancor.ts?5460:5714)) it is on this line: const viewToken = tokens.find(x => compareString(x.id, fee.id))!;`

**Comment by Velua** [comment]
> i still get random errors of TypeError: Cannot read property 'id' of undefined" ethBancor.ts?5460:5714)) it is on this line: const viewToken = tokens.find(x => compareString(x.id, fee.id))!; 

Are we still concerned about those? If so did you add that filter thing I suggested?

> i still get limit errors on localhost with alchemy errors.js?490f:28 Uncaught (in promise) Error: Returned error: Your app has exceeded its concurrent requests capacity. If you have retries enabled, you can safely ignore this message. If not, check out https://docs.alchemyapi.io/documentation/rate-limits. Feel free to ask further questions in the community discord channel: https://discord.com/invite/A39JVCM. Have a great day!

Looking at the code it looks like we're firing off all those converter checks at once, make sure we're both chunking them and sending those chunks one by one, sorted by liq depth. 

Also make sure that ```mutation``` method is only called if we know there's something to be recorded. I don't think it's that big of deal in this case but it's always good to spare running Vue's diffing if possible given how clunky the app currently is.

File: src/store/modules/swap/ethBancor.ts [hunk 1]
```diff
@@ -563,6 +563,64 @@ const determineConverterType = (
...
 };
 
+const getHistoricFees = async (
+  id: string,
+  converterAddress: string,
+  network: EthNetworks
+): Promise<PreviousPoolFee[]> => {
+  const w3 = getWeb3(network);
+  const contract = buildV28ConverterContract(converterAddress, w3);
+  const { blockHoursAgo, currentBlock } = await blockNumberHoursAgo(
```
Comment (Velua) [inline src/store/modules/swap/ethBancor.ts:573]: Pass the blocks as a param to this ```getHistoricFees``` function, need not call on a per converter basis.

File: src/store/modules/swap/ethBancor.ts [hunk 2]
```diff
@@ -563,6 +563,64 @@ const determineConverterType = (
...
+      fromBlock: startBlock,
+      toBlock: endblock
+    };
+
+    const events = await contract.getPastEvents("ConversionFeeUpdate", options);
+
+    history.push(
+      ...events.map(e => ({
+        id,
+        oldDecFee: e.returnValues["_prevFee"] / (10000 * 100),
```
Comment (Velua) [inline src/store/modules/swap/ethBancor.ts:594]: I've previously defined a ```decToPpm``` function, define a ```ppmToDec``` as well. ```(10000 * 100)``` won't be immediately obvious to a new reader as to what this is.

**Review by Velua — CHANGES_REQUESTED** [review]
Looks like we're on the right track, good stuff. Just a few things touch up.

File: src/store/modules/swap/ethBancor.ts [hunk 3]
```diff
@@ -5609,7 +5693,7 @@ export class EthBancorModule
 
           const decFee =
             findPreviousPoolFee(
-              previousPoolFees,
+              this.previousPoolFeesArr,
```
[the remark this answers is not in the document]
Comment (H34D) [reply src/store/modules/swap/ethBancor.ts:5713]: thanks for the hint, i changed that

File: src/store/modules/swap/ethBancor.ts [hunk 4]
```diff
@@ -6235,6 +6303,26 @@ export class EthBancorModule
...
+    const relaysByLiqDepth = this.relays.sort(sortByLiqDepth);
+
+    const relaysList = sortAlongSide(
+      pools,
+      relay => relay.id,
+      relaysByLiqDepth.map(relay => relay.id)
+    );
+
+    const historicFees: PreviousPoolFee[] = (
+      await Promise.all(
```
[the remark this answers is not in the document]
Comment (H34D) [reply src/store/modules/swap/ethBancor.ts:6333]: the good thing is that checkFees is called in batches of 30s pools already, so i did both. I use chunk to limit parallelism

File: src/store/modules/swap/ethBancor.ts [hunk 5]
```diff
@@ -6245,6 +6333,8 @@ export class EthBancorModule
     const allPools = [...pools];
     const allReserveFeeds = [...reserveFeeds];
 
+    void this.checkFees(allPools);
```
[the remark this answers is not in the document]
Comment (H34D) [reply src/store/modules/swap/ethBancor.ts:6353]: done

File: src/store/modules/swap/ethBancor.ts (hunk 1 above)
Comment (H34D) [reply src/store/modules/swap/ethBancor.ts:573]: done

File: src/store/modules/swap/ethBancor.ts (hunk 2 above)
Comment (H34D) [reply src/store/modules/swap/ethBancor.ts:594]: done

**Review by Velua — APPROVED** [review]

**Pull request merged**